client/src/pages/challenges.js

```diff
@@ -30,95 +33,129 @@ export default withStyles({
           problems
         })
       })
+
+    getPrivateSolves()
+      .then(data => {
+        const solveIDs = []
+        data.forEach(solve => {
+          solveIDs.push(solve.id)
+        })
+        this.setState({
+          solveIDs: solveIDs
+        })
+      })
   }
 
   submitFlag = id => e => {
     e.preventDefault()
 
     submitFlag(id, this.state.values[id])
       .then(({ error }) => {
-        const nxt = this.state.errors
-        nxt[id] = error
+        if (error === undefined) {
+          // Flag was submitted successfully
+          util.toasts.useToast().add('Flag successfully submitted!')
+          const newSolveIDs = this.state.solveIDs.concat(id)
+          this.setState({ solveIDs: newSolveIDs })
+        } else {
+          const nxt = this.state.errors
+          nxt[id] = error
 
-        this.setState({
-          errors: nxt
-        })
+          this.setState({
+            errors: nxt
+          })
+        }
       })
   }
 
-  render ({ classes }, { problems, values, errors }) {
+  handleShowSolvesCheckbox = () => {
+    this.setState({
+      showSolved: !this.state.showSolved
+    })
+  }
+
+  renderProblem = (classes, problem, values, errors) => {
```

Comment:
Probably out of scope for this PR, but we should refactor this into a new component later